src/main/java/ch/njol/skript/SkriptCommandTabCompleter.java

```diff
@@ -0,0 +1,109 @@
+/**
+ *   This file is part of Skript.
+ *
+ *  Skript is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  Skript is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Skript.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright Peter Güttinger, SkriptLang team and contributors
+ */
+package ch.njol.skript;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import org.bukkit.command.Command;
+import org.bukkit.command.CommandSender;
+import org.bukkit.command.TabCompleter;
+
+import ch.njol.skript.tests.runner.TestMode;
+import ch.njol.util.StringUtils;
+
+import javax.annotation.Nullable;
+
+public class SkriptCommandTabCompleter implements TabCompleter {
+		
+	@Override
+	@Nullable
+	public List<String> onTabComplete(CommandSender sender, Command command, String alias, String[] args) {
+		ArrayList<String> options = new ArrayList<>();
+		
+		if (!command.getName().equalsIgnoreCase("skript")) {
+			return null;
+		}
+		
+		if (args[0].equalsIgnoreCase("update") && args.length < 3) {
+			options.add("check");
+			options.add("changes");
+			options.add("download");
+		} else if (args[0].matches("(?i)(reload|disable|enable)")) {
```

Comment:
does spigot automatically check for permissions before calling the tabcompleter? if not, we should be careful because this could leak the names and directory structure of the user's scripts.

Reply:
It does, tested and https://hub.spigotmc.org/stash/projects/SPIGOT/repos/bukkit/browse/src/main/java/org/bukkit/command/SimpleCommandMap.java#220
![image](https://user-images.githubusercontent.com/29547183/140518316-ba63e9ea-1d6b-467c-83a4-c48c7d62e807.png)